src/ast/ast.h

```diff
@@ -16,6 +16,12 @@ class Visitor;
 
 #define DEFINE_ACCEPT void accept(Visitor &v) override;
 
+#define DEFINE_CLONE(T)                                                        \
+  T *clone()                                                                   \
```

Comment:
I feel `clone()` indicates deepcopy, but may it's just me.

Reply:
What would you prefer? I wanted to go with copy and deepcopy, but if you think shallowcopy and deepcopy is better im happy with that too

Reply:
I think copy and deepcopy is fine.

Reply:
but this function does not copy pointers and all pointers become nullptr. So copy_without_pointers()? it's a bit longer though...

Reply:
Maybe:

- copy(), shallow copy, copies everything by value and points to the same subtree as  orig did.
- deepcopy(), creates a copy of the whole subtree.  
- leafcopy(), like shallow copy but not copying the pointers to other nodes, effectively turning this into a leaf node.

Reply:
Or nullcopy

Reply:
+1 for leafcopy

Reply:
done :)